File: CountCompleteTreeNodes.cpp
/*
 * Time Limit Exceeded
 */
#include<iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int countNodes(TreeNode* root) {
        if(root == NULL) return 0;
        return countNodes(root->left) + countNodes(root->right) + 1;
    }
};

int main(){
    TreeNode *a = new TreeNode(1);
    TreeNode *b = new TreeNode(2);
    TreeNode *c = new TreeNode(3);
    TreeNode *d = new TreeNode(4);
    TreeNode *e = new TreeNode(5);
    TreeNode *f = new TreeNode(6);
    
    a->left = b;
    a->right = c;
    b->left = d;
    b->right = e;
    c->left = f;
    
    Solution *solution = new Solution();
    cout<<solution->countNodes(a);
    
    return 0;
}

File: CountCompleteTreeNodes.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        
class Solution:
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        deepl = 0
        deepr = 0
        cur = root
        while cur:
            deepl += 1
            cur = cur.left
        cur = root
        while cur:
            deepr += 1
            cur = cur.right
        if deepl == deepr:
            return (1 << deepl) - 1
        else:
            return self.countNodes(root.left) + self.countNodes(root.right) + 1
            
if __name__ == '__main__':
    a = TreeNode(1)
    b = TreeNode(2)
    c = TreeNode(3)
    d = TreeNode(4)
    e = TreeNode(5)
    f = TreeNode(6)
    
    a.left = b
    a.right = c
    b.left = d
    b.right = e
    c.left = f
    
    solution = Solution()
    print(solution.countNodes(a))
